lib/evmone/eof.hpp

```diff
@@ -136,16 +137,25 @@ enum class EOFValidationError
     too_many_container_sections,
     invalid_container_section_index,
     eofcreate_with_truncated_container,
+    ambiguous_container_kind,
+    incompatible_container_kind,
 
     impossible,
 };
 
+enum class ContainerKind : uint8_t
+{
+    initcode_container,
+    runtime_container,
+    initcode_runtime_container,
+};
+
 /// Determines the EOF version of the container by inspecting container's EOF prefix.
 /// If the prefix is missing or invalid, 0 is returned meaning legacy code.
 [[nodiscard]] uint8_t get_eof_version(bytes_view container) noexcept;
 
 /// Validates whether given container is a valid EOF according to the rules of given revision.
-[[nodiscard]] EVMC_EXPORT EOFValidationError validate_eof(
+[[nodiscard]] EVMC_EXPORT std::variant<EOFValidationError, ContainerKind> validate_eof(
```

Comment:
Not sure it's the best idea how to return ContainerKind from validation, and it leads to a big diff in tests etc.

Reply:
Another idea is to not touch this and add a variant like `validate_eof_and _get_container_kind` only for the cases where it's needed.

Reply:
Or optional pointer paramter to output `container_kind`, but it feels too C-like.

Reply:
I opted to adding a special function (with an ugly name), because it would be needed only in very special context of TXCREATE and creation transactions.